### ISSUE TITLE
Is there any way to add texture to element ?

### ISSUE THREAD

**Issue by OmarKhairyHosny** [xBimTeam/XbimWebUI#251]
I am trying to add texture to element but nothing works
this the created IFC file

ISO-10303-21;
HEADER;
FILE_DESCRIPTION ((''), '2;1');
FILE_NAME ('', '2019-09-25T12:11:13', (''), (''), 'Processor version 5.1.0.0', 'Xbim.IO.MemoryModel', '');
FILE_SCHEMA (('IFC4'));
ENDSEC;
DATA;
#1=IFCPROJECT('34wqFLOJvAQPCjBEyQftJ_',#2,'HelloWall',$,$,$,$,(#19,#22),#7);
#2=IFCOWNERHISTORY(#5,#6,$,.ADDED.,1569413474,$,$,0);
#3=IFCPERSON($,'Team','xBIM',$,$,$,$,$);
#4=IFCORGANIZATION($,'xBimTeam',$,$,$);
#5=IFCPERSONANDORGANIZATION(#3,#4,$);
#6=IFCAPPLICATION(#4,'1.0','Hello Wall Application','HelloWall.exe');
#7=IFCUNITASSIGNMENT((#8,#9,#10,#11,#12,#13,#14,#15,#16));
#8=IFCSIUNIT(*,.LENGTHUNIT.,.MILLI.,.METRE.);
#9=IFCSIUNIT(*,.AREAUNIT.,$,.SQUARE_METRE.);
#10=IFCSIUNIT(*,.VOLUMEUNIT.,$,.CUBIC_METRE.);
#11=IFCSIUNIT(*,.SOLIDANGLEUNIT.,$,.STERADIAN.);
#12=IFCSIUNIT(*,.PLANEANGLEUNIT.,$,.RADIAN.);
#13=IFCSIUNIT(*,.MASSUNIT.,$,.GRAM.);
#14=IFCSIUNIT(*,.TIMEUNIT.,$,.SECOND.);
#15=IFCSIUNIT(*,.THERMODYNAMICTEMPERATUREUNIT.,$,.DEGREE_CELSIUS.);
#16=IFCSIUNIT(*,.LUMINOUSINTENSITYUNIT.,$,.LUMEN.);
#17=IFCCARTESIANPOINT((0.,0.,0.));
#18=IFCAXIS2PLACEMENT3D(#17,$,$);
#19=IFCGEOMETRICREPRESENTATIONCONTEXT('Building Model','Model',3,1.E-05,#18,$);
#20=IFCCARTESIANPOINT((0.,0.));
#21=IFCAXIS2PLACEMENT2D(#20,$);
#22=IFCGEOMETRICREPRESENTATIONCONTEXT('Building Plan View','Plan',2,1.E-05,#21,$);
#23=IFCBUILDING('1aemYeGW9BEexMUtxSpLSj',#2,'Default Building',$,$,#24,$,$,.ELEMENT.,$,$,$);
#24=IFCLOCALPLACEMENT($,#25);
#25=IFCAXIS2PLACEMENT3D(#26,$,$);
#26=IFCCARTESIANPOINT((0.,0.,0.));
#27=IFCRELAGGREGATES('0qcVnCnm9E7AelwBKWjDSb',#2,$,$,#1,(#23));
#28=IFCWALLSTANDARDCASE('3f$euyf9r6XPkcjAd6nQPP',#2,'A Standard rectangular wall',$,$,#44,#43,$,$);
#29=IFCRECTANGLEPROFILEDEF(.AREA.,$,#31,300.,4000.);
#30=IFCCARTESIANPOINT((0.,0.));
#31=IFCAXIS2PLACEMENT2D(#30,$);
#32=IFCEXTRUDEDAREASOLID(#29,#35,#33,2400.);
#33=IFCDIRECTION((0.,0.,1.));
#34=IFCCARTESIANPOINT((0.,0.,0.));
#35=IFCAXIS2PLACEMENT3D(#34,$,$);
#36=IFCSTYLEDITEM($,(#41),$);
#37=IFCCOLOURRGB($,0.3,0.6,0.8);
#38=IFCIMAGETEXTURE(.T.,.T.,$,$,$,'\X2\202A\X0\E:\\Upwork\\logo.PNG');
#39=IFCSURFACESTYLEWITHTEXTURES((#38));
#40=IFCSURFACESTYLE($,.BOTH.,(#39));
#41=IFCPRESENTATIONSTYLEASSIGNMENT((#40));
#42=IFCSHAPEREPRESENTATION(#19,'Body','SweptSolid',(#32));
#43=IFCPRODUCTDEFINITIONSHAPE($,$,(#42));
#44=IFCLOCALPLACEMENT($,#45);
#45=IFCAXIS2PLACEMENT3D(#34,#46,$);
#46=IFCDIRECTION((0.,0.,1.));
#47=IFCMATERIAL('some material',$,$);
#48=IFCSTYLEDREPRESENTATION(#19,$,$,(#36));
#49=IFCMATERIALDEFINITIONREPRESENTATION($,$,(#48),#47);
#50=IFCMATERIALLAYERSETUSAGE(#51,.AXIS2.,.NEGATIVE.,150.,$);
#51=IFCMATERIALLAYERSET((#52),$,$);
#52=IFCMATERIALLAYER(#47,10.,$,$,$,$,$);
#53=IFCRELASSOCIATESMATERIAL('0DXAZjxNvDkQZ9vpRWfNw7',#2,$,$,(#28),#50);
#54=IFCRELCONTAINEDINSPATIALSTRUCTURE('1s42r$VTP8PQAHT0fFJbQM',#2,$,$,(#28),#23);
ENDSEC;
END-ISO-10303-21;

**Comment by andyward** [comment]
If you look at that IFC you'll see that line 38 is defining an image texture to be accessed from `E:\Upwork\logo.PNG` 

Even if you managed to feed this through to a wexbim file (it doesn't yet support textures AFAIK) it's hard to understand how this would work in a cloud environment, when logo.png lives on a non-internet facing machine (presumably the local machine of whoever exported the IFC.)

I guess if the path were a valid public URL the WebGL client shaders could be extended to apply available textures. Alternatively an IfcBlobTexture could be used to embed the texture in the IFC, although I've never seen this in practice. Again the web client would need to be updated and the Wexbim format extended to support this.

Bottom line is textures seem to be poorly supported in IFC universally. We could add some support for it in the webUI but it's a wider systemic issue to solve. E.g. Who's outputting IFCs with textures? Where are those texture files hosted if external?

**Comment by OmarKhairyHosny** [comment]
Hi,
Thanks for your reply
How can I extend the shaders to apply textures

**Comment by andyward** [comment]
This won't be a simple thing to implement unless you're pretty familiar with WebGL. You may want to take a look at something like this generic webgl tutorial first: https://webglfundamentals.org/webgl/lessons/webgl-3d-textures.html

I'm not an expert myself, but, to implement you'll want to look at the `viewer.draw()` function https://github.com/xBimTeam/XbimWebUI/blob/3997572703400ab2ae9a123702c7d8e557f066f9/Xbim.WeXplorer/Viewer/viewer.ts#L1458

**Comment by xujunpro2** [comment]
We are trying to do this, we developed the revit plugin, exported the ifc file, and exported the associated texture file, and defined an associated mapping table for the {component ID-texture file} that needs to be texture mapped.
